### PR TITLE
feat: ✨ add Snyk Code in-editor annotations [ROAD-668]

### PR DESCRIPTION
**Pull request by bastiandoetsch**
- added for the following languages & dependencies

Dependency - Language
org.jetbrains.plugins.go - go
org.jetbrains.kotlin - kotlin
com.intellij.java - JAVA
com.intellij.modules.python - Python
com.jetbrains.php - PHP
JavaScript - JavaScript
com.intellij.modules.rider - dotnet

Also for 'TEXT', 'HTML' and 'ANY' languages

### CODE REVIEW

**Comment by bastiandoetsch** [comment]
PHP: 
image

File: src/main/resources/META-INF/optional/io.snyk.snyk-intellij-plugin-withJavaScript.xml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<idea-plugin>
+  <extensions defaultExtensionNs="com.intellij">
+    <externalAnnotator language="JavaScript" implementationClass="snyk.code.annotator.SnykCodeAnnotator"/>
```
Comment (michelkaporin) [inline src/main/resources/META-INF/optional/io.snyk.snyk-intellij-plugin-withJavaScript.xml:3]: Can we add TypeScript as well?

File: src/integTest/kotlin/snyk/code/annotator/SnykCodeAnnotatorTest.kt [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    fun `test textRange for maven pom`() {
+        val issue = createSnykCodeResultWithIssues()[0]
+        val expectedStart = 132
+        val expectedEnd = 203
+        val expectedRange = TextRange(expectedStart, expectedEnd)
+
+        val actualRange = cut.textRange(psiFile, issue.ranges[0])
+
+        assertEquals(expectedRange, actualRange)
+    }
```
Comment (michelkaporin) [inline src/integTest/kotlin/snyk/code/annotator/SnykCodeAnnotatorTest.kt:68]: Can we test some boundary cases for text ranges? E.g. start being at the most beginning of files, etc.

File: src/main/resources/META-INF/optional/io.snyk.snyk-intellij-plugin-withJavaScript.xml (hunk 1 above)
Comment (bastiandoetsch) [reply src/main/resources/META-INF/optional/io.snyk.snyk-intellij-plugin-withJavaScript.xml:3]: Love the question - I THINK, it is covered by JavaScript. But this kind of information is next to impossible to find in the documentation. That said, let's try it out.

File: src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    fun annotationMessage(suggestion: SuggestionForFile): String {
+        return buildString {
+            val title = if (suggestion.title.isNotBlank()) " " + suggestion.title + "." else ""
+            append("Snyk Code:$title ${suggestion.message}")
+        }
+    }
+
+    fun textRange(psiFile: PsiFile, snykCodeRange: MyTextRange): TextRange {
+        val document = psiFile.viewProvider.document ?: return TextRange.EMPTY_RANGE
+        val lineOffSet = document.getLineStartOffset(snykCodeRange.startRow - 1) + snykCodeRange.startCol
```
Comment (michelkaporin) [inline src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt:57]: Don't know if it's 0 or 1-base indexing here - just verifying if that could be prone to IndexOutOfBoundsException?

Comment (ArtsiomCh) [reply src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt:57]: `if (line < 0 || document.lineCount <= line) return TextRange.EMPTY_RANGE`

File: src/main/resources/META-INF/optional/io.snyk.snyk-intellij-plugin-withJavaScript.xml (hunk 1 above)
Comment (ArtsiomCh) [reply src/main/resources/META-INF/optional/io.snyk.snyk-intellij-plugin-withJavaScript.xml:3]: `language="TypeScript"` for .ts files
`language="TypeScript  JSX"` for .tsx files
see [here ](https://github.com/DeepCodeAI/jetbrains-plugin/blob/e228308541c5d0bfb13156f219095859af33f683/src/main/resources/META-INF/plugin.xml#L33)what I did find experimentally ~one year ago

**Comment by ArtsiomCh** [comment]
@bastiandoetsch Rebase on current master needed before review

File: src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import com.intellij.lang.annotation.HighlightSeverity
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiFile
+import io.snyk.plugin.snykcode.core.AnalysisData
+
+class SnykCodeAnnotator : ExternalAnnotator<PsiFile, Unit>() {
+    // overrides needed for the Annotator to invoke apply(). We don't do anything here
+    override fun collectInformation(file: PsiFile): PsiFile = file
+    override fun doAnnotate(collectedInfo: PsiFile?): Unit = Unit
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt:16]: LOG.debug("Calling doAnnotate on ${collectedInfo?.name}")
        val psiFile = collectedInfo ?: return
        val document = PsiDocumentManager.getInstance(psiFile.project).getDocument(psiFile) ?: return
        ApplicationManager.getApplication().invokeLater {
            FileDocumentManager.getInstance().saveDocument(document)
        }

**Comment by bastiandoetsch** [comment]
Rebased...

File: src/main/kotlin/snyk/container/YAMLImageExtractor.kt [hunk 5]
```diff
@@ -5,10 +5,10 @@ import com.intellij.openapi.project.Project
 import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.psi.PsiFile
 import io.snyk.plugin.findPsiFileIgnoringExceptions
-import java.util.stream.Collectors
 
 object YAMLImageExtractor {
     private val logger = Logger.getInstance(javaClass.name)
+    private val imageRegEx = "(?<=image:)(\\s*[a-zA-Z0-9./\\-_]+)(:)?([a-zA-Z0-9./\\-_]+)?".toRegex()
```
Comment (michelkaporin) [inline src/main/kotlin/snyk/container/YAMLImageExtractor.kt:11]: Did you craft regex yourself or copied from somewhere? If latter, let's link the source place for maintenance purposes 🙏

File: src/main/resources/META-INF/plugin.xml [hunk 6]
```diff
@@ -6,21 +6,26 @@
...
+  <depends optional="true" config-file="optional/withYAML.xml">org.jetbrains.plugins.yaml</depends>
+  <depends optional="true" config-file="optional/withXML.xml">com.intellij.modules.xml</depends>
 
   <extensions defaultExtensionNs="com.intellij">
     <toolWindow id="Snyk"
                 anchor="bottom"
                 icon="SnykIcons.TOOL_WINDOW"
                 factoryClass="io.snyk.plugin.ui.toolwindow.SnykToolWindowFactory"/>
 
+    <!--    <externalAnnotator language="TEXT" implementationClass="snyk.code.annotator.SnykCodeAnnotator"/>-->
```
Comment (michelkaporin) [inline src/main/resources/META-INF/plugin.xml:26]: Do we still need these comments?

File: src/main/kotlin/snyk/oss/annotator/OSSNpmAnnotator.kt [hunk 7]
```diff
@@ -44,35 +46,31 @@ class OSSNpmAnnotator : OSSBaseAnnotator() {
...
+                this.foundTextRange = TextRange(element.textRange.startOffset, value.textRange.endOffset)
+                return
             }
             super.visitElement(element)
         }
 
         private fun isSearchedDependency(element: PsiElement): Boolean {
-            return element is JsonStringLiteral && element.value == packageName
+            return element !is PsiComment && element !is PsiWhiteSpace
+                && element is JsonStringLiteral
```
Comment (michelkaporin) [inline src/main/kotlin/snyk/oss/annotator/OSSNpmAnnotator.kt:72]: Let's fix detekt warns 🥇

File: src/main/kotlin/snyk/oss/annotator/OSSNpmAnnotator.kt [hunk 8]
```diff
@@ -44,35 +46,31 @@ class OSSNpmAnnotator : OSSBaseAnnotator() {
...
-            fixVersion
-        }
+        val upgrade = getUpgradeProposal(vulnerability, remediation)
+        val split = upgrade?.upgradeTo?.split("@") ?: return ""
+        return "\"${split[0]}\": \"${split[1]}\""
     }
 
     internal class NpmRecursiveVisitor(
-        private val packageName: String, private val artifactVersion: String
+        private val artifactName: String, private val artifactVersion: String
```
Comment (michelkaporin) [inline src/main/kotlin/snyk/oss/annotator/OSSNpmAnnotator.kt:55]: Let's fix warns

**Review by michelkaporin — APPROVED** [review]
LGTM, commented on some changes to address before merging though 👍

File: src/main/kotlin/snyk/container/YAMLImageExtractor.kt (hunk 5 above)
Comment (bastiandoetsch) [reply src/main/kotlin/snyk/container/YAMLImageExtractor.kt:11]: Crafted myself...

**Comment by bastiandoetsch** [comment]
Good calls @michelkaporin , cleaned up the detekt warnings and comments - as the plugin is always crashing, i had it disabled.

File: src/main/kotlin/snyk/container/YAMLImageExtractor.kt (hunk 5 above)
Comment (michelkaporin) [reply src/main/kotlin/snyk/container/YAMLImageExtractor.kt:11]: Poor you.. 😅

File: src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+            val title = if (suggestion.title.isNotBlank()) " " + suggestion.title + "." else ""
+            append("Snyk Code:$title ${suggestion.message}")
+        }
+    }
+
+    fun textRange(psiFile: PsiFile, snykCodeRange: MyTextRange): TextRange {
+        val document = psiFile.viewProvider.document ?: return TextRange.EMPTY_RANGE
+        // ensure we stay within doc boundaries
+        // Snyk Code is 1-based, Document is 0-based indexes
+        val startRow = max(0, snykCodeRange.startRow - 1)
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt:64]: check against `document.lineCount - 1` needed

File: src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        val endRow = max(startRow, min(document.lineCount - 1, snykCodeRange.endRow - 1))
+        val startCol = max(0, snykCodeRange.startCol)
+        val endCol = max(startCol, min(document.getLineEndOffset(endRow), snykCodeRange.endCol))
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt:67]: endCol could be on the other line and less than startCol
also comparison of offset (relates to the beginning of file) with Column (relates to the beginning of Line) is wrong

File: src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        val suggestions = getIssuesForFile(psiFile)
+
+        suggestions.forEach { suggestionForFile ->
+            val severity = severity(suggestionForFile)
+            suggestionForFile.ranges.forEach {
+                val textRange = textRange(psiFile, it)
+                if (!textRange.isEmpty) {
+                    val annotationBuilder =
+                        holder.newAnnotation(severity, annotationMessage(suggestionForFile)).range(textRange)
+                    annotationBuilder.create()
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt:33]: minor: what's the purpose of `annotationBuilder` here?

File: src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        // ensure we stay within doc boundaries
+        // Snyk Code is 1-based, Document is 0-based indexes
+        val startRow = max(0, snykCodeRange.startRow - 1)
+        val endRow = max(startRow, min(document.lineCount - 1, snykCodeRange.endRow - 1))
+        val startCol = max(0, snykCodeRange.startCol)
+        val endCol = max(startCol, min(document.getLineEndOffset(endRow), snykCodeRange.endCol))
+
+        val lineOffSet = min(document.getLineStartOffset(startRow) + startCol, document.textLength - 1)
+        val lineOffSetEnd = max(lineOffSet, min(document.getLineStartOffset(endRow) + endCol, document.textLength - 1))
+        return TextRange.create(lineOffSet, lineOffSetEnd)
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt:71]: to make it clear: 
Rows are counting from the begging of file
Cols are counting from the begging of row/line
Offset is always counting from the begging of file (if `from LineStart/Element etc.` is not specified explicitly in Document method javadoc)

File: src/main/kotlin/snyk/common/AnnotatorCommon.kt [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+import snyk.oss.annotator.AnnotatorHelper
+
+object AnnotatorCommon {
+    fun prepareAnnotate(psiFile: PsiFile?) {
+        val filePath = psiFile?.virtualFile?.path ?: return
+
+        if (AnnotatorHelper.isFileSupported(filePath)) {
+            val document = psiFile.viewProvider.document ?: return
+            ApplicationManager.getApplication().invokeLater {
+                FileDocumentManager.getInstance().saveDocument(document)
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/common/AnnotatorCommon.kt:15]: minor: logger.debug with the file name could useful here

File: src/main/kotlin/snyk/container/YAMLImageExtractor.kt (hunk 5 above)
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/container/YAMLImageExtractor.kt:11]: why not just take everything after `image:` until whitespace? Do we really need to rely on image name format here?

File: src/main/kotlin/snyk/container/YAMLImageExtractor.kt [hunk 14]
```diff
@@ -29,16 +29,19 @@ object YAMLImageExtractor {
...
+        } else {
+            ""
+        }
     }
 
     fun extractFromFile(file: VirtualFile, project: Project): Set<KubernetesWorkloadImage> {
         val psiFile = findPsiFileIgnoringExceptions(file, project)
-        if (psiFile == null || file.isDirectory || !file.isValid) return emptySet()
+        if (psiFile == null || file.isDirectory || !file.isValid) {
+            logger.warn("no psi file found for ${file.path}")
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/container/YAMLImageExtractor.kt:42]: will produce too many warns for folders/dirs
PS also for any deleted/in-memory file, I guess

File: src/main/kotlin/snyk/container/YAMLImageExtractor.kt [hunk 15]
```diff
@@ -50,7 +53,7 @@ object YAMLImageExtractor {
         isKubernetesFileExtension(psiFile) && isKubernetesFileContent(getFileLines(psiFile))
 
     private fun isKubernetesFileExtension(psiFile: PsiFile): Boolean =
-        psiFile.virtualFile.extension == "yaml"
+        psiFile.virtualFile.extension == "yaml" || psiFile.virtualFile.extension == "yml"
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/container/YAMLImageExtractor.kt:56]: does k8s files could be "yml"? can't find any related docs...

File: src/main/kotlin/snyk/container/YAMLImageExtractor.kt (hunk 5 above)
Comment (bastiandoetsch) [reply src/main/kotlin/snyk/container/YAMLImageExtractor.kt:11]: Yes, you could have a comment after the image name.
